### PR TITLE
Update plugin docs

### LINKED ISSUE #73327: Problems with the list of plugins in plugins/plugins.rst
##### SUMMARY

When you run the `ansible-galaxy collection init NAMESPACE.COLLECTION` command, it creates a sample collection directory layout.  In the plugins folder there is a README.md file which lists the directories for plugins which ansible recognizes.  It reads as follows:

> Here is an example directory of the majority of plugins currently supported by Ansible:
>
> ```
> └── plugins
>     ├── action
>     ├── become
>     ├── cache
>     ├── callback
>     ├── cliconf
>     ├── connection
>     ├── filter
>     ├── httpapi
>     ├── inventory
>     ├── lookup
>     ├── module_utils
>     ├── modules
>     ├── netconf
>     ├── shell
>     ├── strategy
>     ├── terminal
>     ├── test
>     └── vars
> ```
>.
> A full list of plugin types can be found at [Working With Plugins](https://docs.ansible.com/ansible/devel/plugins/plugins.html).

The list of plugin types on https://docs.ansible.com/ansible/devel/plugins/plugins.html has some problems, however:
* It is missing terminal plugins so it's actually a shorter list than what's in the README.md
* doc_fragments plugins are missing from the list too
* modules and module_utils aren't in that list.  Traditionally Ansible doesn't group those with the other plugins but if we're pointing people at that page who are populating a collection's `plugins/` directory, we probably should mention those as well.
* Filter plugins are listed as `Using filters to manipulate data`.  Someone relatively new to Ansible might not know that those are called "filter plugins" (and thus the directory should be `plugins/'filters`) from that description.  They might think they are `data plugins` or something else instead.
* Test plugins are listed as simply `Tests`.  The entry should probably be `Test Plugins`
* The last entry in that list is `Rejecting modules`.  That page is not about a specific type of plugin but a page which tells how to configure Ansible to exclude certain **modules**.  An area of the documentation for configuration is probably a better location for this.  If we want it on this page, it could have its own paragraph or be part of the seealso block.

##### ISSUE TYPE
- Documentation Report

##### COMPONENT NAME
docs/docsite/rst/plugins/plugins.rst

##### ANSIBLE VERSION

```paste below
devel
```

##### CONFIGURATION

```paste below
N/A
```

##### OS / ENVIRONMENT

N/A

### PR DESCRIPTION
**Pull request by acozine**
##### SUMMARY
Closes #73327
Supersedes #74769 

Adds a page for every plugin type, adds cross-linking to multiple related topics, edits existing plugin pages for style, etc.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
docs.ansible.com

### CODE REVIEW

File: docs/docsite/rst/plugins/callback.rst [hunk 1]
```diff
@@ -1,23 +1,20 @@
...
 ================
 
 .. contents::
    :local:
    :depth: 2
 
-Callback plugins enable adding new behaviors to Ansible when responding to events.
-By default, callback plugins control most of the output you see when running the command line programs,
-but can also be used to add additional output, integrate with other tools and marshall the events to a storage backend.
+Callback plugins enable adding new behaviors to Ansible when responding to events. By default, callback plugins control most of the output you see when running the command line programs, but can also be used to add additional output, integrate with other tools and marshall the events to a storage backend. If necessary, you can :ref:`create custom callback plugins <developing_callbacks>`.
```
Comment (Akasurde) [inline docs/docsite/rst/plugins/callback.rst]:
```suggestion
Callback plugins enable adding new behaviors to Ansible when responding to events. By default, callback plugins control most of the output you see when running the command-line programs, but can also be used to add additional output, integrate with other tools, and marshall the events to a storage backend. If necessary, you can :ref:`create custom callback plugins <developing_callbacks>`.
```

Comment (acozine) [reply docs/docsite/rst/plugins/callback.rst]: Good catch, thanks @Akasurde.

File: docs/docsite/rst/plugins/docs_fragment.rst [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+.. seealso::
+
+   :ref:`developing_modules_general`
+       An introduction to creating Ansible modules
+   :ref:`developing_collections`
+       An guide to creating Ansible collections
+   `User Mailing List <https://groups.google.com/group/ansible-devel>`_
+       Have a question?  Stop by the google group!
+   `irc.freenode.net <http://irc.freenode.net>`_
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/docs_fragment.rst]: This should point to Libera.

File: docs/docsite/rst/plugins/filter.rst [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+       Inventory plugins
+   :ref:`callback_plugins`
+       Callback plugins
+   :ref:`test_plugins`
+      Test plugins
+   :ref:`lookup_plugins`
+       Lookup plugins
+   `User Mailing List <https://groups.google.com/group/ansible-devel>`_
+       Have a question?  Stop by the google group!
+   `irc.freenode.net <http://irc.freenode.net>`_
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/filter.rst]: Should point to Libera.

File: docs/docsite/rst/plugins/module_util.rst [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+.. seealso::
+
+   :ref:`developing_modules_general`
+       An introduction to creating Ansible modules
+   :ref:`developing_collections`
+       An guide to creating Ansible collections
+   `User Mailing List <https://groups.google.com/group/ansible-devel>`_
+       Have a question?  Stop by the google group!
+   `irc.freenode.net <http://irc.freenode.net>`_
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/module_util.rst]: Libera.

File: docs/docsite/rst/plugins/module_util.rst [hunk 5]
```diff
@@ -0,0 +1,35 @@
+.. _module_util_plugins:
+
+Module utilities
+================
+
+.. contents::
+   :local:
+   :depth: 2
+
+Module utilities contain shared code used by multiple modules. You can write :ref:`custom module utilities <developing_module_utilities>`.
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/module_util.rst]: Code from the `module_utils` directory can (and often is) also used by other plugin kinds, not only modules.

File: docs/docsite/rst/plugins/terminal.rst [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+       Connection plugins
+   :ref:`cliconf_plugins`
+       Network cliconf plugins
+   :ref:`httpapi_plugins`
+       Network httpapi plugins
+   :ref:`netconf_plugins`
+       Network netconf plugins
+   `User Mailing List <https://groups.google.com/group/ansible-devel>`_
+       Have a question?  Stop by the google group!
+   `irc.freenode.net <http://irc.freenode.net>`_
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/terminal.rst]: Libera.

File: docs/docsite/rst/plugins/test.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+       Using conditional statements
+   :ref:`filter_plugins`
+       Filter plugins
+   :ref:`playbooks_filters`
+       Using filters
+   :ref:`lookup_plugins`
+       Lookup plugins
+   `User Mailing List <https://groups.google.com/group/ansible-devel>`_
+       Have a question?  Stop by the google group!
+   `irc.freenode.net <http://irc.freenode.net>`_
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/test.rst]: Libera.

File: docs/docsite/rst/plugins/action.rst [hunk 8]
```diff
@@ -1,16 +1,15 @@
...
+Action plugins
 ==============
 
 .. contents::
    :local:
    :depth: 2
 
-Action plugins act in conjunction with :ref:`modules <working_with_modules>` to execute the actions required by playbook tasks.
-They usually execute automatically in the background doing prerequisite work before modules execute.
+Action plugins act in conjunction with :ref:`modules <working_with_modules>` to execute the actions required by playbook tasks. They usually execute automatically in the background doing prerequisite work before modules execute.
```
Comment (tadeboro) [inline docs/docsite/rst/plugins/action.rst:10]: Maybe we should add something along the lines of "action plugin is the control node part of a module"? This should help explain why action plugins share their documentation with the module.

**Review by tadeboro — CHANGES_REQUESTED** [review]
I found a few Freenode references in the docs, but apart from that, things do look OK from my side.

File: docs/docsite/rst/plugins/module_util.rst (hunk 5 above)
Comment (acozine) [reply docs/docsite/rst/plugins/module_util.rst]: ah, good catch, thanks

File: docs/docsite/rst/plugins/docs_fragment.rst (hunk 2 above)
Comment (acozine) [reply docs/docsite/rst/plugins/docs_fragment.rst]: well spotted! I've fixed those in 6f2b1c71c73aa833220fc996bb108ef5a520f63f

File: docs/docsite/rst/plugins/callback.rst (hunk 1 above)
Comment (samccann) [inline docs/docsite/rst/plugins/callback.rst]:
```suggestion
Callback plugins enable adding new behaviors to Ansible when responding to events. By default, callback plugins control most of the output you see when running the command line programs, but can also be used to add additional output, integrate with other tools and marshal the events to a storage backend. If necessary, you can :ref:`create custom callback plugins <developing_callbacks>`.
```

**Review by samccann — COMMENTED** [review]
one nit

**Review by tadeboro — APPROVED** [review]
Thank you, @acozine, for preparing this PR.

**Pull request merged**